File: app/annotations.py
"""标注存储（annotations.json）：高亮 + 书签 CRUD、导出。

数据模型：
  {"version":1, "books": {"<book_id>": {
      "highlights": [{"id","chapter_index","start_offset","end_offset",
                      "text","color","note","created_at","updated_at"}],
      "bookmarks":  [{"id","chapter_index","offset","text","created_at"}]
  }}}

偏移为章节内折叠纯文本字符坐标（见 app/text.py），与前端 TextPos 对齐。
"""
import json
import threading
import uuid
from pathlib import Path
from typing import Optional

from .storage import atomic_write_json, load_json_file, now_iso

HL_COLORS = ("yellow", "green", "blue", "pink", "purple", "cyan")


class AnnotationStore:
    """per-book 高亮与书签。线程安全，原子写。"""

    def __init__(self, file: Path):
        self._file = file
        self._lock = threading.RLock()
        self._books: dict = {}

    def load(self) -> None:
        data = load_json_file(self._file)
        self._books = data.get("books", {}) if data else {}

    def save(self) -> None:
        with self._lock:
            atomic_write_json(self._file, {"version": 1, "books": self._books})

    def _book(self, book_id: str) -> dict:
        b = self._books.setdefault(book_id, {"highlights": [], "bookmarks": []})
        b.setdefault("highlights", [])
        b.setdefault("bookmarks", [])
        return b

    # ---------- 读取 ----------

    def get_highlights(self, book_id: str) -> list[dict]:
        with self._lock:
            return [dict(h) for h in self._book(book_id)["highlights"]]

    def get_bookmarks(self, book_id: str) -> list[dict]:
        with self._lock:
            return [dict(b) for b in self._book(book_id)["bookmarks"]]

    def get_all(self, book_id: str) -> dict:
        return {"highlights": self.get_highlights(book_id), "bookmarks": self.get_bookmarks(book_id)}

    # ---------- 高亮 ----------

    def add_highlight(
        self,
        book_id: str,
        chapter_index: int,
        start_offset: int,
        end_offset: int,
        text: str,
        color: str = "yellow",
        note: str = "",
    ) -> dict:
        start_offset = max(0, int(start_offset))
        end_offset = max(0, int(end_offset))
        if end_offset <= start_offset:
            raise ValueError("高亮区间无效")
        if color not in HL_COLORS:
            color = "yellow"
        now = now_iso()
        ann = {
            "id": uuid.uuid4().hex[:12],
            "chapter_index": max(0, int(chapter_index)),
            "start_offset": start_offset,
            "end_offset": end_offset,
            "text": (text or "")[:2000],
            "color": color,
            "note": (note or "")[:5000],
            "created_at": now,
            "updated_at": now,
        }
        with self._lock:
            self._book(book_id)["highlights"].append(ann)
            self.save()
        return dict(ann)

    def update_annotation(self, book_id: str, ann_id: str, patch: dict) -> Optional[dict]:
        with self._lock:
            for h in self._book(book_id)["highlights"]:
                if h["id"] == ann_id:
                    if "note" in patch:
                        h["note"] = (patch.get("note") or "")[:5000]
                    if "color" in patch and patch.get("color") in HL_COLORS:
                        h["color"] = patch["color"]
                    if "text" in patch:
                        h["text"] = (patch.get("text") or "")[:2000]
                    h["updated_at"] = now_iso()
                    self.save()
                    return dict(h)
            return None

    def delete_annotation(self, book_id: str, ann_id: str) -> bool:
        with self._lock:
            b = self._book(book_id)
            before = len(b["highlights"])
            b["highlights"] = [h for h in b["highlights"] if h["id"] != ann_id]
            if len(b["highlights"]) != before:
                self.save()
                return True
            return False

    # ---------- 书签 ----------

    def add_bookmark(self, book_id: str, chapter_index: int, offset: int, text: str) -> dict:
        bm = {
            "id": uuid.uuid4().hex[:12],
            "chapter_index": max(0, int(chapter_index)),
            "offset": max(0, int(offset)),
            "text": (text or "")[:200],
            "created_at": now_iso(),
        }
        with self._lock:
            self._book(book_id)["bookmarks"].append(bm)
            self.save()
        return dict(bm)

    def delete_bookmark(self, book_id: str, bm_id: str) -> bool:
        with self._lock:
            b = self._book(book_id)
            before = len(b["bookmarks"])
            b["bookmarks"] = [x for x in b["bookmarks"] if x["id"] != bm_id]
            if len(b["bookmarks"]) != before:
                self.save()
                return True
            return False

    def remove_book(self, book_id: str) -> None:
        with self._lock:
            if book_id in self._books:
                del self._books[book_id]
                self.save()

    # ---------- 导出 ----------

    def export(self, book_id: str, fmt: str, book_title: str, chapter_title_fn) -> str:
        """导出为 markdown 或 json。chapter_title_fn(index)->str。"""
        with self._lock:
            b = self._book(book_id)
            highlights = [dict(h) for h in b["highlights"]]
            bookmarks = [dict(x) for x in b["bookmarks"]]
        if fmt == "json":
            return json.dumps(
                {"book": book_title, "highlights": highlights, "bookmarks": bookmarks},
                ensure_ascii=False,
                indent=2,
            )
        # markdown：按章分组
        lines = [f"# {book_title} 标注导出", ""]
        by_chapter: dict[int, list] = {}
        for h in highlights:
            by_chapter.setdefault(h["chapter_index"], []).append(("hl", h))
        for bm in bookmarks:
            by_chapter.setdefault(bm["chapter_index"], []).append(("bm", bm))
        for ci in sorted(by_chapter):
            lines.append(f"## {chapter_title_fn(ci)}")
            for kind, item in by_chapter[ci]:
                if kind == "hl":
                    lines.append(f"> {item['text']}")
                    if item.get("note"):
                        lines.append(f"笔记：{item['note']}")
                else:
                    lines.append(f"🔖 {item['text']}")
                lines.append("")
        return "\n".join(lines)

File: android/app/src/main/java/io/github/gighi947/ankeshelf/data/Annotations.kt
package io.github.gighi947.ankeshelf.data

import kotlinx.serialization.Serializable
import java.io.File
import java.util.UUID
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

val HL_COLORS: List<String> = listOf("yellow", "green", "blue", "pink", "purple", "cyan")

@Serializable
data class Highlight(
    val id: String,
    val chapter_index: Int,
    val start_offset: Int,
    val end_offset: Int,
    val text: String,
    val color: String = "yellow",
    val note: String = "",
    val created_at: String = "",
    val updated_at: String = "",
)

@Serializable
data class Bookmark(
    val id: String,
    val chapter_index: Int,
    val offset: Int,
    val text: String,
    val created_at: String = "",
)

@Serializable
data class AnnotationBook(
    val highlights: List<Highlight> = emptyList(),
    val bookmarks: List<Bookmark> = emptyList(),
)

@Serializable
data class AnnotationsFile(
    val version: Int = 1,
    val books: Map<String, AnnotationBook> = emptyMap(),
)

/** 标注存储（annotations.json）：高亮 + 书签 CRUD、导出；text_offset 坐标。 */
class AnnotationStore(private val file: File) {

    private val lock = ReentrantLock()
    private val writeGuard = StoreWriteGuard()
    private var books: MutableMap<String, AnnotationBook> = mutableMapOf()

    fun load(): List<StoreLoadIssue> {
        val (loaded, issue) = loadGuarded(file, writeGuard) { AnnotationsFile() }
        books = loaded.books.toMutableMap()
        return listOfNotNull(issue)
    }

    fun save() {
        if (writeGuard.writeBlocked()) {
            logWarn("AnkeShelf", "annotations.json 读取失败过，跳过写入以保护原文件")
            return
        }
        val snapshot = lock.withLock { books.toMap() }
        atomicWriteJson(
            file,
            Shelf.json.encodeToString(AnnotationsFile.serializer(), AnnotationsFile(books = snapshot)),
        )
    }

    private fun book(bookId: String): AnnotationBook =
        books.getOrPut(bookId) { AnnotationBook() }

    fun getHighlights(bookId: String): List<Highlight> =
        lock.withLock { book(bookId).highlights }

    fun getBookmarks(bookId: String): List<Bookmark> =
        lock.withLock { book(bookId).bookmarks }

    fun getAll(bookId: String): AnnotationBook =
        lock.withLock { book(bookId) }

    fun addHighlight(
        bookId: String,
        chapterIndex: Int,
        startOffset: Int,
        endOffset: Int,
        text: String,
        color: String = "yellow",
        note: String = "",
    ): Highlight {
        val start = maxOf(0, startOffset)
        val end = maxOf(0, endOffset)
        if (end <= start) throw IllegalArgumentException("高亮区间无效")
        val now = nowIso()
        val ann = Highlight(
            id = UUID.randomUUID().toString().replace("-", "").take(12),
            chapter_index = maxOf(0, chapterIndex),
            start_offset = start,
            end_offset = end,
            text = text.take(2000),
            color = if (color in HL_COLORS) color else "yellow",
            note = note.take(5000),
            created_at = now,
            updated_at = now,
        )
        lock.withLock {
            val b = book(bookId)
            books[bookId] = b.copy(highlights = b.highlights + ann)
            save()
        }
        return ann
    }

    fun updateAnnotation(bookId: String, annId: String, patch: AnnotationPatch): Highlight? =
        lock.withLock {
            val b = book(bookId)
            val idx = b.highlights.indexOfFirst { it.id == annId }
            if (idx < 0) return null
            val old = b.highlights[idx]
            val updated = old.copy(
                note = patch.note?.take(5000) ?: old.note,
                color = patch.color?.takeIf { it in HL_COLORS } ?: old.color,
                text = patch.text?.take(2000) ?: old.text,
                updated_at = nowIso(),
            )
            val list = b.highlights.toMutableList()
            list[idx] = updated
            books[bookId] = b.copy(highlights = list)
            save()
            updated
        }

    fun deleteAnnotation(bookId: String, annId: String): Boolean = lock.withLock {
        val b = book(bookId)
        val list = b.highlights.filterNot { it.id == annId }
        if (list.size != b.highlights.size) {
            books[bookId] = b.copy(highlights = list)
            save()
            true
        } else {
            false
        }
    }

    fun addBookmark(bookId: String, chapterIndex: Int, offset: Int, text: String): Bookmark {
        val bm = Bookmark(
            id = UUID.randomUUID().toString().replace("-", "").take(12),
            chapter_index = maxOf(0, chapterIndex),
            offset = maxOf(0, offset),
            text = text.take(200),
            created_at = nowIso(),
        )
        lock.withLock {
            val b = book(bookId)
            books[bookId] = b.copy(bookmarks = b.bookmarks + bm)
            save()
        }
        return bm
    }

    fun deleteBookmark(bookId: String, bmId: String): Boolean = lock.withLock {
        val b = book(bookId)
        val list = b.bookmarks.filterNot { it.id == bmId }
        if (list.size != b.bookmarks.size) {
            books[bookId] = b.copy(bookmarks = list)
            save()
            true
        } else {
            false
        }
    }

    fun removeBook(bookId: String) {
        lock.withLock {
            if (books.remove(bookId) != null) save()
        }
    }

    /** 导出为 markdown 或 json（chapterTitleFn(index) -> 章节标题）。 */
    fun export(
        bookId: String,
        fmt: String,
        bookTitle: String,
        chapterTitleFn: (Int) -> String,
    ): String {
        val snapshot = lock.withLock {
            val b = book(bookId)
            b.highlights to b.bookmarks
        }
        if (fmt == "json") {
            return Shelf.json.encodeToString(
                ExportJson.serializer(),
                ExportJson(book = bookTitle, highlights = snapshot.first, bookmarks = snapshot.second),
            )
        }
        val lines = mutableListOf("# $bookTitle 标注导出", "")
        val byChapter = sortedMapOf<Int, MutableList<Pair<String, Any>>>()
        snapshot.first.forEach { byChapter.getOrPut(it.chapter_index) { mutableListOf() }.add("hl" to it) }
        snapshot.second.forEach { byChapter.getOrPut(it.chapter_index) { mutableListOf() }.add("bm" to it) }
        for ((ci, items) in byChapter) {
            lines.add("## ${chapterTitleFn(ci)}")
            for ((kind, item) in items) {
                if (kind == "hl") {
                    val h = item as Highlight
                    lines.add("> ${h.text}")
                    if (h.note.isNotEmpty()) lines.add("笔记：${h.note}")
                } else {
                    val bm = item as Bookmark
                    lines.add("🔖 ${bm.text}")
                }
                lines.add("")
            }
        }
        return lines.joinToString("\n")
    }
}

data class AnnotationPatch(
    val note: String? = null,
    val color: String? = null,
    val text: String? = null,
)

@Serializable
private data class ExportJson(
    val book: String,
    val highlights: List<Highlight>,
    val bookmarks: List<Bookmark>,
)
